### PR TITLE
[SDK-3717] Add tree-shakable BaseClient and Rest exports

### LINKED ISSUE #1374: Create tree-shakeable `BaseClient` and `Rest` modules
- Create two classes, `BaseClient` and `Rest` which are available from a separate import path from the main ably-js module
- `BaseClient` optionally takes `Rest` as a second constructor argument, and if so, allows the REST API to be accessed
- Should include a typescript interface for the new import path

### PR DESCRIPTION
**Pull request by owenpearson**
Resolves #1374 

Refactors the existing `Rest` class into `BaseClient` and an encapsulated `Rest` class which can be passed in to the `BaseClient` constructor to allow access to the rest interface.

The tree-shakable modules are currently available as an esm module from 'ably/modules', example usage:
```ts
import { BaseClient, Rest } from 'ably/modules';

const client = new BaseClient(options, { Rest });
```

A new npm script (`npm run modulereport`) is added which displays information about the size of each module in the bundle-report workflow. Currently ([76d2ab0](https://github.com/ably/ably-js/pull/1400/commits/76d2ab0eb9237cc0f0f863a5b0a500b9a5ad50b7)) the output shows:

```
BaseClient: 61.63kb
BaseClient + Rest: 78.56kb
```

These numbers will be lower after we encapsulate further functionality into tree-shakable modules.

### CODE REVIEW

File: src/common/lib/client/realtime.ts [hunk 1]
```diff
@@ -14,17 +14,21 @@ import Platform from 'common/platform';
 import Message from '../types/message';
 
 class Realtime extends Rest {
-  channels: any;
```
Comment (lawrence-forooghian) [inline src/common/lib/client/realtime.ts:17]: I'm not sure I understand why this change is needed here?

**Review by lawrence-forooghian — APPROVED** [review]
I'm approving (with one comment), but as discussed offline, I'm still not sure whether it makes sense to combine the realtime and REST functionality into a single `BaseClient` class. My preference would be to resolve this matter before merging this PR, but if you'd prefer to defer the decision until a bit later then that's cool too 👍

**Comment by jamienewcomb** [comment]
@owenpearson can we remember to link PRs to the jira issue they are fixing please. Happy to have a discussion about why this is required need be.

It can be done by tagging the issue key in the title. see https://github.com/ably/ably-go/pull/596
Or by adding it to the PR description. see https://github.com/ably/ably-java/pull/966

This ask is now reflected in: https://github.com/ably/engineering/blob/main/best-practices/pull-requests.md

File: src/common/lib/client/baseclient.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  _currentFallback: null | {
+    host: string;
+    validUntil: number;
+  };
+  serverTimeOffset: number | null;
+  http: IHttp;
+  auth: Auth;
+  _rest?: Rest;
+
+  constructor(options: ClientOptions | string, modules: ModulesMap) {
```
Comment (lawrence-forooghian) [inline src/common/lib/client/baseclient.ts:31]:
```suggestion
  constructor(options: ClientOptions | string, modules?: ModulesMap) {
```

(given the usage without passing modules in `modules.test.js` and the `if (modules?.Rest)` below)

File: test/browser/modules.test.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import { BaseClient, Rest } from '../../build/modules/index.js';
```
Comment (lawrence-forooghian) [inline test/browser/modules.test.js:1]: Does this need to be added to `test/support/browser_file_list.js`?

File: test/browser/modules.test.js
Comment (lawrence-forooghian) [inline test/browser/modules.test.js:1]: Does this need to be added to `test/support/browser_file_list.js`?

File: src/common/lib/client/defaultrest.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * Preloaded BaseClient with all REST features, used as the default non-treeshakeable Rest export
+ */
+class DefaultRest extends BaseClient {
+  constructor(options: string | ClientOptions) {
+    super(options, {
+      Rest,
+    });
+  }
+
+  static Platform = Platform;
```
Comment (lawrence-forooghian) [inline src/common/lib/client/defaultrest.ts:18]: Shouldn't these static properties all go on `BaseClient`? They're part of the API that it's meant to satisfy.

Comment (lawrence-forooghian) [reply src/common/lib/client/defaultrest.ts:18]: (and `BaseClient` already has a static `Platform` property)

File: test/browser/modules.test.js
Comment (lawrence-forooghian) [reply test/browser/modules.test.js:1]: My mistake — fixed as part of our call

File: src/common/lib/client/realtime.ts (hunk 1 above)
Comment (owenpearson) [reply src/common/lib/client/realtime.ts:17]: tsc errors if you extend a class and override a dynamic accessor to be a static field, even if the return type of the accessor is the same as the static field

File: src/common/lib/client/defaultrest.ts (hunk 4 above)
Comment (owenpearson) [reply src/common/lib/client/defaultrest.ts:18]: i've removed `Platform` here though, nice spot 👍

File: scripts/moduleReport.js
Comment (lawrence-forooghian) [inline scripts/moduleReport.js:1]: Can we update the `format` task to include this file?

**Comment by lawrence-forooghian** [comment]
Replaced by #1417.
